source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc

```diff
@@ -30,9 +32,9 @@ class SimpleLookupContext : public LookupContext {
     cb(std::make_unique<Buffer::OwnedImpl>(&body_[range.begin()], range.length()));
   }
 
-  void getTrailers(LookupTrailersCallback&&) override {
-    // TODO(toddmgreer): Support trailers.
-    NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+  void getTrailers(LookupTrailersCallback&& cb) override {
+    cb(trailers_ ? Http::createHeaderMap<Http::ResponseTrailerMapImpl>(*trailers_)
+                 : Http::ResponseTrailerMapPtr{});
```

Comment:
Do we have a testcase that hits the case where trailers is not present?

This was pre-existing but it'd be good to make sure we also have a case for where headers are not present and so the getHeaders callback gets called with am empty LookupResult.

Also, WDYT of using absl::optional rather than empty-constructed objects for both those cases?

Reply:
jmarantz, I think you're suggesting changing LookupTrailersCallback's parameter from ResponseTrailerMapPtr&& to optional<ResponseTrailerMap>&&. That would cause an unnecessary copy, because these trailers get passed to StreamDecoderFilterCallbacks::encodeTrailers, which requires ResponseTrailerMapPtr&&.

LookupTrailerCallback should take the same parameter type as StreamDecoderFilterCallbacks::encodeTrailers does.

Reply:
The declaration comment of HttpCache::getTrailers should say something like "The cache must call cb with the cached trailers." (Sorry for the omission; feel free to add that if so inclined.) I suggest something like:

void getTrailers(LookupTrailersCallback&& cb) override {
  ASSERT(trailers_);
  cb(Http::createHeaderMap<Http::ResponseTrailerMapImpl>(*trailers_));
}

CacheFilter promises to only call getTrailers if LookupResult::has_trailers_ is true. If getTrailers gets called while trailers_ is null, it's a bug.

Reply:
Thanks Todd -- keeping the type as is SGTM.

There's still my pending question "Do we have a testcase that hits the case where trailers is not present?"

Reply:
Sorry, yes, I have added the requested test case in https://github.com/envoyproxy/envoy/pull/16854/commits/df64f002894465e01a1fb695c14de82e106710e6.